### PR TITLE
Added check for deprecated bitwise operation with float in php8.1.

### PR DESCRIPTION
**Pull request by mad-briller**
closes https://github.com/phpstan/phpstan/issues/7700

### CODE REVIEW

File: src/Rules/Operators/InvalidBinaryOperationRule.php [hunk 1]
```diff
@@ -113,6 +116,30 @@ public function processNode(Node $node, Scope $scope): array
...
+				$node instanceof Node\Expr\BinaryOp\ShiftRight
+				|| $node instanceof Node\Expr\BinaryOp\ShiftLeft
+				|| $node instanceof Node\Expr\BinaryOp\BitwiseOr
+				|| $node instanceof Node\Expr\BinaryOp\BitwiseAnd
+				|| $node instanceof Node\Expr\BinaryOp\BitwiseXor
+			) {
+				$leftType = $scope->getType($node->left);
+				$rightType = $scope->getType($node->right);
+				$floatType = new FloatType();
+				if ($floatType->isSuperTypeOf($leftType)->yes() || $floatType->isSuperTypeOf($rightType)->yes()) {
```
Comment (mad-briller) [inline src/Rules/Operators/InvalidBinaryOperationRule.php:130]: not sure whether these checks should actually be `maybe()` to account for union types etc?

**Comment by mad-briller** [comment]
the issue raised in github actions about `FileReadTrapStreamWrapper` raises a good question, should these checks be disabled for mixed?

**Comment by mad-briller** [comment]
looking at the details of what has been deprecated, it's all implicit float -> int conversions, so perhaps this is a bandaid rather than a real fix
https://php.watch/versions/8.1/deprecate-implicit-conversion-incompatible-float-string

**Comment by ondrejmirtes** [comment]
I think this direction wasn't exactly wrong. We should cover the other scenarios as well, the ones I encountered myself personally were:

1) The ones covered in this PR
2) Floats in array keys
3) Floats on the right side `%` (maybe left too? not sure)

Of course https://php.watch/versions/8.1/deprecate-implicit-conversion-incompatible-float-string has an exhaustive list we can work with.

What I imagine should be done is to create new rules for these scenarios. These rules should apply RuleLevelHelper, and be enabled only with bleedingEdge for now (and report the things only on PHP 8.1+).

The way we work with RuleLevelHelper is like this:

```php
$floatType = new FloatType();
$typeResult = $this->ruleLevelHelper->findTypeToCheck(
    $scope,
    $expr,
    '', // error message about unknown class - can be skipped here
    static fn (Type $type): bool => $floatType->isSuperTypeOf($type)->no(),
);
$type = $typeResult->getType();
if ($type instanceof ErrorType) {
    return [];
}
if ($floatType->isSuperTypeOf($type)->no()) {
    return [];
}

// report error here
```

What this is gonna achieve - a type where we're sure about it being float is going to always be reported, something that's only partially a float (like `int|float`) is going to be reported on level 7, and explicit `mixed` is going to be reported on level 9.

**Pull request closed without merging**